node/overseer/src/lib.rs

```diff
@@ -534,798 +331,104 @@ pub async fn forward_events<P: BlockchainEvents<Block>>(
-				self.collator_protocol_unbounded
-					.unbounded_send(make_packet(signals_received, msg))
-					.map_err(|e| e.into_send_error())
-			},
-			AllMessages::ApprovalDistribution(msg) => {
-				self.approval_distribution_unbounded
-					.unbounded_send(make_packet(signals_received, msg))
-					.map_err(|e| e.into_send_error())
-			},
-			AllMessages::ApprovalVoting(msg) => {
-				self.approval_voting_unbounded
-					.unbounded_send(make_packet(signals_received, msg))
-					.map_err(|e| e.into_send_error())
-			},
-			AllMessages::GossipSupport(msg) => {
-				self.gossip_support_unbounded
-					.unbounded_send(make_packet(signals_received, msg))
-					.map_err(|e| e.into_send_error())
-			},
-			AllMessages::DisputeCoordinator(_) => Ok(()),
-			AllMessages::DisputeParticipation(_) => Ok(()),
-			AllMessages::ChainSelection(_) => Ok(()),
-		};
-
-		if res.is_err() {
-			tracing::debug!(
-				target: LOG_TARGET,
-				"Failed to send a message to another subsystem",
-			);
-		}
-	}
-}
-
-type SubsystemIncomingMessages<M> = stream::Select<
-	metered::MeteredReceiver<MessagePacket<M>>,
-	metered::UnboundedMeteredReceiver<MessagePacket<M>>,
->;
-
-#[derive(Debug, Default, Clone)]
-struct SignalsReceived(Arc<AtomicUsize>);
-
-impl SignalsReceived {
-	fn load(&self) -> usize {
-		self.0.load(atomic::Ordering::SeqCst)
-	}
-
-	fn inc(&self) {
-		self.0.fetch_add(1, atomic::Ordering::SeqCst);
-	}
-}
-
-/// A sender from subsystems to other subsystems.
-#[derive(Debug, Clone)]
-pub struct OverseerSubsystemSender {
-	channels: ChannelsOut,
-	signals_received: SignalsReceived,
-}
-
-#[async_trait::async_trait]
-impl SubsystemSender for OverseerSubsystemSender {
-	async fn send_message(&mut self, msg: AllMessages) {
-		let needed_signals = self.signals_received.load();
-		self.channels.send_and_log_error(needed_signals, msg).await;
-	}
-
-	async fn send_messages<T>(&mut self, msgs: T)
-		where T: IntoIterator<Item = AllMessages> + Send, T::IntoIter: Send
-	{
-		// This can definitely be optimized if necessary.
-		for msg in msgs {
-			self.send_message(msg).await;
-		}
-	}
-
-	fn send_unbounded_message(&mut self, msg: AllMessages) {
-		self.channels.send_unbounded_and_log_error(self.signals_received.load(), msg);
-	}
-}
-
-/// A context type that is given to the [`Subsystem`] upon spawning.
-/// It can be used by [`Subsystem`] to communicate with other [`Subsystem`]s
-/// or to spawn it's [`SubsystemJob`]s.
-///
-/// [`Overseer`]: struct.Overseer.html
-/// [`Subsystem`]: trait.Subsystem.html
-/// [`SubsystemJob`]: trait.SubsystemJob.html
-#[derive(Debug)]
-pub struct OverseerSubsystemContext<M>{
-	signals: metered::MeteredReceiver<OverseerSignal>,
-	messages: SubsystemIncomingMessages<M>,
-	to_subsystems: OverseerSubsystemSender,
-	to_overseer: metered::UnboundedMeteredSender<ToOverseer>,
-	signals_received: SignalsReceived,
-	pending_incoming: Option<(usize, M)>,
-	metrics: Metrics,
-}
+/// The `Overseer` itself.
+#[overlord(
```

Comment:
Nice naming :laughing:

Reply:
Whoops, but I am tempted to leave it as is :sweat_smile: